implement forum

ThomasJuster opened the pull request:
## Description

Fixes #...

## Notes (out of scope, known issues, hints for reviewing code, ...)  (optional)

:warning: To be tested locally. Install https://github.com/France-ioi/AlgoreaForum project
```sh
git clone git@github.com:France-ioi/AlgoreaForum.git
cd AlgoreaForum
# OR
cd path/to/AlgoreaForum
git checkout master
git pull

npm ci
npm start
```

## Test cases

NOTE: The database is in-memory, restarting the server process will flush it. To ensure tests behave correctly, relaunch the server before each test.

Test 1:
1. In a first tab, navigate to [this task](http://localhost:4200/#/activities/by-id/6379723280369399253;path=4702,1625159049301502151,5207993233955027100;attempId=0/details) as demo user
2. In a second tab, navigate to [this task](http://localhost:4200/#/activities/by-id/6379723280369399253;path=4702,1625159049301502151,5207993233955027100;attempId=0/details?watchedGroupId=752024252804317630&watchUser=1) as the usual user watching the demo user
  3. The thread is not opened, I see on tab 1 the "Open" button, and on tab 2 I see nothing

Test 2:
1. In a first tab, navigate to [this task](http://localhost:4200/#/activities/by-id/6379723280369399253;path=4702,1625159049301502151,5207993233955027100;attempId=0/details) as demo user
2. In a second tab, navigate to [this task](http://localhost:4200/#/activities/by-id/6379723280369399253;path=4702,1625159049301502151,5207993233955027100;attempId=0/details?watchedGroupId=752024252804317630&watchUser=1) as the usual user watching the demo user
3. And in tab 1 I click on the "Open" button
4. Then In tab 1, I see the imported history and I can send a message ; in tab 2, I see nothing changed.
5. And I refresh tab 2 (or navigate to another item and back to this one)
6. Then in tab 2, I see thread events and I can send a message.
7. And in tab 1, I send a message
8. Then in both tabs I see the message
9. And in tab 2, I send a message
10. Then in both tabs I see the message
11. And I reload tab 1
12. Then in tab 1, I see thread events including previously sent messages.
13. And I reload tab 2 (make sure you still have the watch-related query parameters!)
14. Then in tab 2, I see thread events including sent messages.
15. And I send again messages in either tab
16. Then both tabs receive it
17. And in tab 1, I close the thread
18. Then in both tabs I cannot send messages
19. And in tab 1, I reopen the thread
20. Then in tab 1, I see thread events including old messages ; in tab 2, I can send messages again.

smadbe commented:
1. I would expect the 2 message streams to contain the same events. But with the follow/unfollow, they do not contain the same list... it even looks strange as the "unfollow" appears in the past... and when the user is back it is even less consistent.
Anyway, from what I remember, the `unfollow` removes 'follow`. If so, probably it is easier not to show them in event list at all. Possibly we could have a list of user who are currently following somewhere else. 

2. When reopening a thread, I would expect users to view the previous messages. If the observed user close the thread, refresh the page and reopen the thread, he does not see the former messages. (however if he re-refreshes, he sees them)

ThomasJuster commented:
> 1. I would expect the 2 message streams to contain the same events. But with the follow/unfollow, they do not contain the same list... it even looks strange as the "unfollow" appears in the past... and when the user is back it is even less consistent.
Anyway, from what I remember, the unfollow removes 'follow`. If so, probably it is easier not to show them in event list at all. Possibly we could have a list of user who are currently following somewhere else.

I don't see how it is a problem, here the display it raw, it's up to the frontend to pick the messages it is interested in to extract messages, followers and so on. To me the inconsistencies are due to a raw display which will change anyway.
Would you expect this work to be done on frontend or backend side? And what is to be done exactly, remove "follow"/"unfollow" events from the list of events we send?

> 2. When reopening a thread, I would expect users to view the previous messages. If the observed user close the thread, refresh the page and reopen the thread, he does not see the former messages. (however if he re-refreshes, he sees them)

What use case do you mention? When watching a user with a closed thread, we should see the previous messages?

smadbe commented:
In test 1 "I see on tab 1 the "Open" button, and on tab 2 I see nothing" -> I do not have an "Open" button and if I click on the thread-related button, nothing happens:
https://user-images.githubusercontent.com/1053150/181029970-4d799bb0-1007-4d34-9a33-214496e2c299.mov

smadbe commented:
Few more remarks:
- if there is an opened thread for observer, the ui should auto show the thread
- haven't we decided that when a user open a thread, he sees his own follow? I don't see it
- if I 1) open a thread with a user, 2) join the thread with an observer, 3) let the user close the thread, 4) let the user reopen the thread... the user just see 3 events: attempt_stated, submission and thread_open... so actually "his UI" does not know that the observer is still following. And btw, why doesn't he see the previous messages? he should

ThomasJuster commented:
Fixed

smadbe commented:
When the observed user open, close, then reopen the thread, I see a follow (in the observer log). However, the oberserved user should have the same behavior of keeping the observation while the thread is closed. In the near future, we may have some observers allowed to open thread, so users should know about it.

smadbe commented:
-  After a refresh of the observer (so I assume a flush of the history locally), I see several "follow" event from the observed user. So it looks like there are several "follow" from the same user in the db which should not be possible. It looks like the UI does not correctly unfollow as it should. 

![Screenshot 2022-07-28 at 18 34 00](https://user-images.githubusercontent.com/1053150/181591017-eb53bd6f-7dba-4fec-b0a7-cccab408d4cb.png)

- I cannot reproduce but after leaving the page open a few minutes, I noticed that the observed user was not receiving the messages anymore. And if he submit something, no error but nothing was sent. Refreshing the page fixed the problem, but still...

- If I watch the observer while I refresh the observed user page, the "unfollow" of the observed user is arriving in the past in the history (at the same time as the previous follow, right?). Is it a normal behavior?

ThomasJuster commented:
> If I watch the observer while I refresh the observed user page, the "unfollow" of the observed user is arriving in the past in the history (at the same time as the previous follow, right?). Is it a normal behavior?

Indeed, I fixed it in AlgoreaForum

> After a refresh of the observer (so I assume a flush of the history locally), I see several "follow" event from the observed user. So it looks like there are several "follow" from the same user in the db which should not be possible. It looks like the UI does not correctly unfollow as it should.

I tried changing the page and refreshing the page, I don't have any trouble now, maybe the unfollow-fix also fixed that part but I doubt it.


> When the observed user opens, closes, then reopens the thread, I see a follow (in the observer log). However, the oberserved user should have the same behavior of keeping the observation while the thread is closed. In the near future, we may have some observers allowed to open thread, so users should know about it.

I don't understand this one. What I did was:
1. Connect as the observed user, open thread.
2. Connect as the observer user
3. As the observed user, close thread then re-open thread
4. Then in the observer user thread I have 2 "follow" event with user id of observed.
That's the issue ?
If so, I just fixed it on AlgoreaForum as well.

smadbe commented:
The last problem is still there. Just having the 2 users arriving on their page, I open the thread on the observed user and show the thread on both sides. I get:

On observed user-side:
```
{ "time": "2021-12-01T09:41:13.000Z", "eventType": "attempt_started", "attemptId": "0" }
{ "time": "2021-12-01T09:41:38.000Z", "eventType": "submission", "attemptId": "0", "answerId": "3700639321047855207", "score": 25, "validated": false }
{ "time": "2022-08-01T12:29:46.301Z", "eventType": "thread_opened", "byUserId": "752024252804317630" }
```

On observer side:
```
{ "time": "2021-12-01T09:41:13.000Z", "eventType": "attempt_started", "attemptId": "0" }
{ "time": "2021-12-01T09:41:38.000Z", "eventType": "submission", "attemptId": "0", "answerId": "3700639321047855207", "score": 25, "validated": false }
{ "time": "2022-08-01T12:29:26.129Z", "eventType": "follow", "userId": "752024252804317630" }
{ "time": "2022-08-01T12:29:37.910Z", "eventType": "follow", "userId": "670968966872011405" }
{ "time": "2022-08-01T12:29:46.301Z", "eventType": "thread_opened", "byUserId": "752024252804317630" }
```

So the observed user does not know the observer follows him. (and btw he does not even know he follows the thread himself)

ThomasJuster commented:
Fixed

smadbe commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
+    this.clearEvents$.next();
+    this.activityLogService.getActivityLog(itemId, watchedGroupId).subscribe({
+      next: history => this.send({ action: 'open-thread', history }),
+      error: () => { /* FIXME: What to do? */ }
```
At least display a red toast. 
As nothing is created in a such case, it seems quite safe the leave the rest as it is.

ThomasJuster replied:
Fixed

smadbe commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
+  }
+
+  unfollow(): void {
+    this.send({ action: 'unfollow' });
```
`unfollow` seems to be called from multiple place. Probably no need to send several of them. 
A "status of the thread" should probably be known/stored in the service so that multiple follow and unfollow are not sent. (with a `throw`  if it is really supposed to be impossible)

ThomasJuster replied:
The fact that a user follows or not does not depend on the thread status anymore, that means: at task init we follow, at task destroy (or window unload) we unfollow.

smadbe commented on src/app/modules/item/components/thread/thread.component.ts:
```diff
@@ -38,21 +57,41 @@ export class ThreadComponent implements OnChanges, OnDestroy {
+    if (!this.itemData || !changes.itemData) return;
+    const previousValue = changes.itemData.previousValue as ItemData | undefined;
+    if (previousValue && this.itemData.route.id !== previousValue.route.id) this.threadService.unfollow();
     this.itemData$.next(this.itemData);
```
1. This line assumes there is only `itemData` as input. That is true now but this may change, so it would be safe to check.

2. Then, don't forget that `itemData` builds up, so it will change while you do not want to change anything to the thread service, so maybe having only the `itemId` as input would make more sense?

3. Also, in order to convince oneself that the follow and unfollow are called as expected, they should be called exactly at the same / from the same trigger. Here `unfollow` is called in `ngOnChanges` while `follow` is called in the `this.itemData$` subscription.

ThomasJuster replied:
I changed to itemId :+1: 
For the point 3, I used the `pairwise` operator to avoid calling `unfollow` on the first visited item.

smadbe reviewed (changes requested):
Ok, now the behavior looks ok.

Here is my first review about the code.

ThomasJuster commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
+    );
+
+    this.status$ = this.incomingMessages$.pipe(
+      map(messages => messages.find(isThreadStatusMessage)?.status),
+      filter(isNotUndefined),
+      distinctUntilChanged(),
```
You would then have a result of type array, which is not what I want :thinking:

smadbe replied:
Ok, sorry.

And there is a guarantee there is only one single status message among the messages?

ThomasJuster replied:
This part of the code is no longer valid since thread-status is now an http service.
To answer your question though: in the array we keep (which contains the newest events), there might be multiple status-related events (open/close) or none (when the last open/close event is older than the last 20 events at init).
1. If there are thread-status-related events in our array, we consider the last one.
2. If there are none, we ask the thread-status http service.
3. If we start with no thread-status-related event in our array and receive one afterwards, case 1 applies.

Anyhow, the thread status is unique at a given time.

ThomasJuster commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -20,21 +35,60 @@ const threadClosedEventDecoder = D.struct({
+});
+
+type ThreadStatusMessage = D.TypeOf<typeof threadStatusDecoder>;
+const isThreadStatusMessage = (message: ThreadMessage): message is ThreadStatusMessage => !isThreadEventMessage(message);
```
I think you're mistakening the messages the backend can send with the actual thread event we store.
The backend does not only send thread events, it also sends the 'unfollow' event and the thread status, both are not stored in dynamodb

smadbe replied:
Ok, got it.
The "has_thread" is a temp thing while AlgoreaBackend does not handle it itself, it is supposed to be a request-response. Probably that's the kind of things that we should have left out of the WS as it makes things messy. You do not handle a response as you handle a stream of something which may arrive at any moment. We will have to deconstruct that sooner or later anyway.

ThomasJuster replied:
ok

ThomasJuster commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
+    this.subscriptions.push(
+      // Every time we receive an event related to thread status, refetch the status.
+      this.newEvents$.pipe(
+        map(events => events.find(event => event.eventType === 'thread_opened' || event.eventType === 'thread_closed')),
```
I am working with a list of events everywhere. I'm only resetting that list upon some conditions

smadbe replied:
got it

ThomasJuster commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
+      this.newEvents$.pipe(
+        map(events => events.find(event => event.eventType === 'thread_opened' || event.eventType === 'thread_closed')),
+        filter(isNotUndefined),
+      ).subscribe(() => this.getStatus()),
```
Because the handler "thread-status" was required in the spec of the backend, to me it means we want to code the feature on the backend instead of the frontend: https://github.com/France-ioi/AlgoreaForum/issues/20

smadbe replied:
And so you consider we'll call the backend each time we receive a thread_opened or thread_closed to know whether the thread is closed or not?

ThomasJuster replied:
If we infer the thread status on the frontend from the received thread events (which is fine by me), then to me there's no point in creating a 'thread-status' service on any backend. Should I do that?

ThomasJuster commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
+    this.newEvents$ = this.incomingMessages$.pipe(map(messages => messages.filter(isThreadEventMessage)));
+
+    this.events$ = this.clearEvents$.pipe(
+      startWith(undefined),
+      switchMap(() => this.newEvents$.pipe(scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]))),
```
In practice I don't need a seed. If I add a seed, it will an empty array, I'll need to cast it which adds noise in my opinion

smadbe replied:
That's not really my point.  It is to use the `scan` without `seed` as it is a valid usage.
I meant that if the array is [ A, B ] and there is a "clean", nothing will be emitting while there is no new events.

ThomasJuster replied:
What's a clean event?
My test to make sure the behavior is ok was:
```js
const { Subject, timer, scan } = require('rxjs');

const events = new Subject();
timer(1000).subscribe(() => events.next([ 1, 2 ]));
timer(2000).subscribe(() => events.next([ 3, 4 ]));

events
  .pipe(scan((prev, next) => [ ...prev, ...next ]))
  .subscribe(events => console.info(events));
// emits [1,2] after 1s
// emits [1, 2, 3, 4] after 2s
```
"nothing will be emitted while there is no new events." -> did you expect to have only one emission of `[1, 2, 3, 4]` after 2s?

smadbe replied:
> What's a clean event?

"clear" event, sorry.

No, I mean that when `this.clearEvents$` emits, nothing is emitted (so to `this.events$`) while there is no emission on `this.newEvents$`.

ThomasJuster replied:
Got it, I added a `startWith([])` :+1: 
Fixed

ThomasJuster commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
+    this.events$ = this.clearEvents$.pipe(
+      startWith(undefined),
+      switchMap(() => this.newEvents$.pipe(scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]))),
+      map(messages => messages.sort((a, b) => a.time.valueOf() - b.time.valueOf())), // sort by date ascending
```
If 2 messages have the same timestamp, it means they are the same, and that should not happen.

smadbe replied:
In your code until L3 it is "events" not "message".  (is the naming change on purpose btw?) So a "follow" and "open thread" at the same time would not "means they are the same".
So I can imagine there may be 2 events occurring at the end moment.. just because you do not control what you get from activities... for instance having a score change event and a validation event at the same time.

ThomasJuster replied:
> in your code until L3 it is "events" not "message". (is the naming change on purpose btw?)

The naming is on purpose and will change since the `thread-status` action will become an http endpoint and won't be sent as a message anymore. When this work will be done, the frontend will only deal with events, which are in this context thread events. (Side note: although "unfollow" is not stored, it remains a thread event).

> So a "follow" and "open thread" at the same time would not "means they are the same".

This cannot happen due to dynamodb storage: 2 thread events - that are stored - cannot have the same pk+time couple since this couple is unique in dynamodb. If we were to add 2 entries with the same pk+time, the last entry would override the other. This issue is handled on the backend except for the thread events imported from history. Can you check in the DB if there are a lot of logs having the same itemId+userId+datetime values? (datetime precision is ms).

smadbe replied:
For the message/events thing: 
- is there a clear definition of what is an event and a message (and if one includes the other one) somewhere is file? I remember you explain me what I cannot remember the difference anymore.
- so the code above, we work with events on line 139, then work with "messages" on line 140 and then store the results in "event$"... so that's is correct?

> This issue is handled on the backend except for the thread events imported from history. Can you check in the DB if there are a lot of logs having the same itemId+userId+datetime values? (datetime precision is ms).

There is no such thing as logs in the DB, it is computed logs. But nothing prevent them to have a same time so it will happen, so it has to be handled.

ThomasJuster replied:
> For the message/events thing

I changed the thread-status to an http service instead of socket, so from now on the frontend only deals with events, no more "message/events" distinction to make. Every socket message is a thread event.

---

For the activity logs ok, I'll add milliseconds to duplicates on the backend side.

ThomasJuster replied:
> For the activity logs ok, I'll add milliseconds to duplicates on the backend side.

Fixed on the backend, you can pull master

ThomasJuster commented on src/app/modules/item/services/threads.service.ts:
```diff
@@ -44,37 +98,115 @@ interface TokenData {
 
   init(tokenData: TokenData): void {
     this.tokenData = tokenData;
+    this.clearEvents$.next(); // when re-initializing the service, clear event list.
```
I can remove it by using a ReplaySubject. Fixed

smadbe commented on src/app/modules/item/components/thread/thread.component.ts:
```diff
@@ -11,25 +10,45 @@ import { ThreadService } from '../../services/threads.service';
+        this.widgetOpened = status === 'opened';
+      }),
+
+    this.itemId$.pipe(pairwise()).subscribe(() => this.threadService.unfollow()),
```
You use pairwise jut to skip the first event? 🤨 Just use `skip(1)`.

ThomasJuster replied:
Indeed :monocle_face:, fixed

smadbe commented:
A few last comment/questions.

smadbe reviewed (approved):
I think that's good now. Thanks!
